components/mdns/mdns.c

```diff
@@ -6797,4 +6926,604 @@ void mdns_debug_packet(const uint8_t *data, size_t len)
+}
+
+esp_err_t mdns_browse_delete(const char *service, const char *proto)
+{
+    mdns_browse_t *browse = NULL;
+
+    if (!_mdns_server || _str_null_or_empty(service) != _str_null_or_empty(proto)) {
+        return ESP_FAIL;
+    }
+
+    browse = _mdns_browse_init(service, proto, NULL);
+    if (!browse) {
+        return ESP_ERR_NO_MEM;
+    }
+
+    if (_mdns_send_browse_action(ACTION_BROWSE_END, browse)) {
+        _mdns_browse_free(browse);
+        return ESP_ERR_NO_MEM;
+    }
+    return ESP_OK;
+}
+
+/**
+ * @brief  Mark browse as finished, remove and free it from browse chain
+ */
+static void _mdns_browse_finish(mdns_browse_t *browse)
+{
+    browse->state = BROWSE_OFF;
+    mdns_browse_t *b = _mdns_server->browse;
+    mdns_browse_t *target_free = NULL;
+    while (b) {
+        if (strlen(b->service) == strlen(browse->service) && memcmp(b->service, browse->service, strlen(b->service)) == 0 &&
+                strlen(b->proto) == strlen(browse->proto) && memcmp(b->proto, browse->proto, strlen(b->proto)) == 0) {
+            target_free = b;
+            b = b->next;
+            queueDetach(mdns_browse_t, _mdns_server->browse, target_free);
+            _mdns_browse_free(target_free);
+        } else {
+            b = b->next;
+        }
+    }
+    _mdns_browse_free(browse);
+}
+
+/**
+ * @brief  Add new browse to the browse chain
+ */
+static void _mdns_browse_add(mdns_browse_t *browse)
+{
+    browse->state = BROWSE_RUNNING;
+    mdns_browse_t *queue = _mdns_server->browse;
+    bool found = false;
+    // looking for this browse in active browses
+    while (queue) {
+        if (strlen(queue->service) == strlen(browse->service) && memcmp(queue->service, browse->service, strlen(queue->service)) == 0 &&
+                strlen(queue->proto) == strlen(browse->proto) && memcmp(queue->proto, browse->proto, strlen(queue->proto)) == 0) {
+            found = true;
+            break;
+        }
+        queue = queue->next;
+    }
+    if (!found) {
+        browse->next = _mdns_server->browse;
+        _mdns_server->browse = browse;
+    }
+    _mdns_browse_send(browse);
+    if (found) {
+        _mdns_browse_free(browse);
+    }
+}
+
+/**
+ * @brief  Send PTR query packet to all available interfaces for browsing.
+ */
+static void _mdns_browse_send(mdns_browse_t *browse)
+{
+    // Using search once for sending the PTR query
+    mdns_search_once_t search = {0};
+
+    search.instance = NULL;
+    search.service = browse->service;
+    search.proto = browse->proto;
+    search.type = MDNS_TYPE_PTR;
+    search.unicast = false;
+    search.result = NULL;
+    search.next = NULL;
+
+    uint8_t i, j;
+    for (i = 0; i < MDNS_MAX_INTERFACES; i++) {
+        for (j = 0; j < MDNS_IP_PROTOCOL_MAX; j++) {
+            _mdns_search_send_pcb(&search, (mdns_if_t)i, (mdns_ip_protocol_t)j);
+        }
+    }
+}
+
+static esp_err_t _mdns_browse_sync_result_find_or_add_result(mdns_browse_sync_t *sync_browse, mdns_result_t *r)
```

Comment:
could we find a better name? `_mdns_browse_sync_result_find_or_add_result()` doesn't sound very descriptive, maybe something just `add_browse_result()` and comment that if the result's already there we don't have to add it...